lib/web/terminal.go

```diff
@@ -284,12 +289,43 @@ func (t *TerminalHandler) makeClient(ws *websocket.Conn) (*client.TeleportClient
 	tc.OnShellCreated = func(s *ssh.Session, c *ssh.Client, _ io.ReadWriteCloser) (bool, error) {
 		t.sshSession = s
 		t.windowChange(&t.params.Term)
+
+		go t.startPingLoop(ws)
+
 		return false, nil
 	}
 
 	return tc, nil
 }
 
+// startPingLoop starts a loop that will continuously send a ping frame through the websocket
+// to prevent the connection between web client and teleport proxy from becoming idle.
+// Interval is determined by the keep_alive_interval config set by user (or default).
+// Loop will terminate when there is an error sending ping frame or when terminal session is closed.
+func (t *TerminalHandler) startPingLoop(ws *websocket.Conn) {
+	// Define our own marshal func to just return a ping payload type.
+	codec := websocket.Codec{Marshal: func(v interface{}) (data []byte, payloadType byte, err error) {
+		return nil, websocket.PingFrame, nil
+	}}
+
+	t.log.Debugf("Starting websocket ping loop with interval %v.", t.params.KeepAliveInterval)
+	tickerCh := time.NewTicker(t.params.KeepAliveInterval)
+	defer tickerCh.Stop()
+
+	for {
+		select {
+		case <-tickerCh.C:
+			if err := codec.Send(ws, nil); err != nil {
+				t.log.Errorf("Unable to send ping frame to web client: %v.", err)
+				return
```

Comment:
Should `ws` get closed here to terminate all server-side state?

Reply:
I am glad you pointed this out. I wasn't sure whether to close (in this case, I will close the session which will cause the ws to close). I figured that if there was a problem with the ws, it would be handled elsewhere and this loop can just terminate. Should I explicitly handle it here?

Reply:
If keepalive fails we should definitely be shutting things down, but it looks like `websocket.Conn` probably isn't safe for concurrent calls to `Close()`.  One option is to cancel the terminal context (`t.terminalCancel()`), though I'd double-check that this does actually trigger shutdown (some of our client code isn't context-aware).  Another option would be to protect `websocket.Conn.Close()` with a `sync.Once`, though closing by context cancellation is probably preferable.

Reply:
Bumping this since PR has been re-opened.  Looks like we still need shutdown trigger here.

Reply:
revised

Reply:
> If keepalive fails we should definitely be shutting things down, 

@fspmarshall why? if it fails just exist the routine. This should not have any effect on actual connection.